### ISSUE TITLE
Type inference (?) not working for combination of tagged unions

### ISSUE THREAD

**Issue by jeffijoe** [Microsoft/TypeScript#12692]
**TypeScript Version:**  2.2.0-dev.20161130

**Code**

```ts
export type CouponDuration =
  'ONCE' |
  'FOREVER'

export interface ICoupon {
  id: string
  friendlyName?: string
  description?: string
  maxTotalRedemptions: number
  dateExpires?: Date
  duration: CouponDuration
}

export interface IAmountOffCoupon extends ICoupon {
  offType: 'AMOUNT_OFF'
  amountOff: number
}

export interface IPercentOffCoupon extends ICoupon {
  offType: 'PERCENT_OFF'
  percentOff: number
}

export interface IRegularCoupon extends ICoupon {
  couponType: 'REGULAR'
}

export interface ISpecialDiscountCoupon extends ICoupon {
  couponType: 'SPECIAL_DISCOUNT'
}

export interface IReferralCodeCoupon extends ICoupon {
  couponType: 'REFERRAL_CODE'
  referrerUserId: string
}

// Combination
export type Coupon = (IAmountOffCoupon | IPercentOffCoupon)
  & (IRegularCoupon | ISpecialDiscountCoupon | IReferralCodeCoupon)

// Error here
const c: Coupon = {
  id: 'JEFF17HA',
  couponType: 'REFERRAL_CODE',
  referrerUserId: 'abcd1234',
  offType: 'AMOUNT_OFF',
  amountOff: 20
}
```

**Expected behavior:**

The `c` variable should not emit a compiler error.

**Actual behavior:**

`c` variable emits the following error:

```
error TS2322: Type '{ id: string; couponType: "REFERRAL_CODE"; referrerUserId: string; offType: "AMOUNT_OFF"; amountO...' is not assignable to type 'Coupon'.
  Type '{ id: string; couponType: "REFERRAL_CODE"; referrerUserId: string; offType: "AMOUNT_OFF"; amountO...' is not assignable to type 'IPercentOffCoupon & IReferralCodeCoupon'.
    Type '{ id: string; couponType: "REFERRAL_CODE"; referrerUserId: string; offType: "AMOUNT_OFF"; amountO...' is not assignable to type 'IPercentOffCoupon'.
      Types of property 'offType' are incompatible.
        Type '"AMOUNT_OFF"' is not assignable to type '"PERCENT_OFF"'.
```

However casting it fixes the problem:

```ts
const c: Coupon = <IReferralCodeCoupon & IAmountOffCoupon> {
  id: 'JEFF17HA',
  couponType: 'REFERRAL_CODE',
  referrerUserId: 'abcd1234',
  offType: 'AMOUNT_OFF',
  amountOff: 20
}
```
